### DISCUSSION TITLE
如何优雅地升级版本

### DISCUSSION THREAD

**Discussion by sampsonye** [go-atomci/atomci#57] [Q&A] [answered]
# 起因
最近的迭代过程中如果存在数据库或者配置变更，我们很难做到比较`丝滑`地升级
* 针对数据库变更，我们目前只能放到deploy的`init.sh`脚本中维护
* 针对配置变更(#55 )我们没有更好地办法，只能在文档中多加提示

# 思考
针对数据库变更，是否有比较好的Migration组件自动同步变更情况？另，变更发生时，数据迁移如何操作呢？
针对配置变更，是否可以考虑如下几个可能性：
* 【方案一】除必要配置外，其他配置均放到数据库配置表中
* 【方案二】实现类似于数据库Migration工具，自动帮我们增量补齐默认配置
* 【方案三】维护一个`app.config.template`文件，默认最新配置，用于用户参考
* 【方案四】支持配置拆分（类似于nginx的conf.d）不同类别配置可以单独文件（灵活性可能较差）

**Comment by fanhousanbu** [comment]
Gorm 自带的 migrate方法是否能满足？

**Reply by colynn** [reply]
支持方案二， 引入 migration工具， 但是目前我们使用主orm不是 Gorm,   可以引入类似 https://github.com/golang-migrate/migrate   ， 或是再做下其他的调研。

Harbor的开源也是基于beego框架的，引入的migrate框架也是 golang-migrate/migrate

**Reply by sampsonye** [reply]
我们需要对`数据库变更(架构变更和数据迁移)`和`配置变更`输出独立的方案
社区有数据库相关的migration工具，针对配置变更的migration工具有推荐吗

**Reply by fanhousanbu** [reply]
pr #58 也会导致同样的问题；因为之前已经引用了casbin，会自动生成casbin_rule表，所以会引起升级失败；需要手动删除casbin_rule表

**Reply by colynn** [reply]
> 我们需要对`数据库变更(架构变更和数据迁移)`和`配置变更`输出独立的方案 社区有数据库相关的migration工具，针对配置变更的migration工具有推荐吗

这个工具 https://github.com/golang-migrate/migrate

**Comment by colynn** [comment, marked as answer]
总结：两个待办项
1. 新增 app.config.template 的示例配置文件；
2. 调研引入 migrate 工具，比如 https://github.com/golang-migrate/migrate

**Reply by fanhousanbu** [reply]
添加示例：app.config.template，除了文件本身，构建时也需要自动发布吗？

**Reply by colynn** [reply]
docker-compose的tgz 自动更新还是有些其他依赖，逻辑上也是可以
